Rearrange bit fields of JSC::UnlinkedFunctionExecutable for Windows

fujii opened the pull request:
#### 2aa03fba6a7507b4089350c669402ec9f73efc6b
<pre>
Rearrange bit fields of JSC::UnlinkedFunctionExecutable for Windows
<a href="https://bugs.webkit.org/show_bug.cgi?id=255571">https://bugs.webkit.org/show_bug.cgi?id=255571</a>

Reviewed by Yusuke Suzuki.

Clang-cl reported the following static_assert failure for Windows
port. Use uint16_t and uint8_t to pack the bit fields.

&gt; bytecode/UnlinkedFunctionExecutable.h(327,1): error: static assertion failed due to requirement &apos;sizeof(JSC::UnlinkedFunctionExecutable) &lt;= 96&apos;: UnlinkedFunctionExecutable needs to be small
&gt; bytecode/UnlinkedFunctionExecutable.h(327,50): note: expression evaluates to &apos;104 &lt;= 96&apos;

* Source/JavaScriptCore/bytecode/UnlinkedFunctionExecutable.h:

Canonical link: <a href="https://commits.webkit.org/263149@main">https://commits.webkit.org/263149@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/d407f1576318c260c8a94c66a39a7b68497ab227

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/3760 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/3845 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/3953 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/5189 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/4030 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/3923 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/3850 "Built successfully") | [❌ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/3261 "layout-tests (failure)") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/3804 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/4020 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/3389 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/5022 "Built successfully") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/39/builds/1525 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/3361 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/3737 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/3082 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/3338 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/3420 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/4799 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/3532 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/3810 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/3089 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/3834 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/3337 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/3361 "Passed tests") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/965 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/919 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/3376 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/24/builds/3926 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/3615 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/3/builds/1078 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

Constellation commented on Source/JavaScriptCore/bytecode/UnlinkedFunctionExecutable.h:
```diff
@@ -280,13 +280,13 @@ class UnlinkedFunctionExecutable final : public JSCell {
+    uint8_t m_implementationVisibility : bitWidthOfImplementationVisibility;
+    uint8_t m_lexicalScopeFeatures : bitWidthOfLexicalScopeFeatures;
+    uint8_t m_functionMode : 2; // FunctionMode
+    uint8_t m_derivedContextType: 2;
```
Let's add space before `:` while it is not newly changed part.

fujii replied:
Fixed.

Constellation commented on Source/JavaScriptCore/bytecode/UnlinkedFunctionExecutable.h:
```diff
@@ -280,13 +280,13 @@ class UnlinkedFunctionExecutable final : public JSCell {
     unsigned m_privateBrandRequirement : 1;
-    unsigned m_features : bitWidthOfCodeFeatures;
-    unsigned m_constructorKind : 2;
+    uint16_t m_features : bitWidthOfCodeFeatures;
```
Let's use `CodeFeatures` instead of `uint16_t`.

fujii replied:
Fixed.

Constellation reviewed (approved):
r=me

fujii committed:
Rearrange bit fields of JSC::UnlinkedFunctionExecutable for Windows
https://bugs.webkit.org/show_bug.cgi?id=255571

Reviewed by Yusuke Suzuki.

Clang-cl reported the following static_assert failure for Windows
port. Use uint16_t and uint8_t to pack the bit fields.

> bytecode/UnlinkedFunctionExecutable.h(327,1): error: static assertion failed due to requirement 'sizeof(JSC::UnlinkedFunctionExecutable) <= 96': UnlinkedFunctionExecutable needs to be small
> bytecode/UnlinkedFunctionExecutable.h(327,50): note: expression evaluates to '104 <= 96'

* Source/JavaScriptCore/bytecode/UnlinkedFunctionExecutable.h:

Canonical link: https://commits.webkit.org/263149@main
